libcontainer/process_linux.go

```diff
@@ -81,9 +84,40 @@ func (p *setnsProcess) signal(sig os.Signal) error {
 
 func (p *setnsProcess) start() (retErr error) {
 	defer p.messageSockPair.parent.Close()
+
 	// get the "before" value of oom kill count
 	oom, _ := p.manager.OOMKillCount()
-	err := p.cmd.Start()
+
+	// When greater or equal to zero, it will set a temporary single CPU
+	// affinity before cgroup cpuset transition, this handles a corner
+	// case when joining a container having all the processes running
+	// exclusively on isolated CPU cores to force the kernel to schedule
+	// runc process on the first CPU core within the cgroups cpuset.
+	// The introduction of the kernel commit 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76
+	// has affected this deterministic scheduling behavior by distributing
+	// tasks across CPU cores within the cgroups cpuset. Some intensive
+	// real-time application are relying on this deterministic behavior
+	// and use the first CPU core to run a slow thread while other CPU
+	// cores are fully used by real-time threads with SCHED_FIFO policy.
+	// Such applications prevent runc process from joining a container
+	// when the runc process is randomly scheduled on a CPU core owned
+	// by a real-time thread.
+	cpuAffinity := -1
+
+	if len(p.manager.GetPaths()) > 0 {
+		// get the target container cgroup
+		if cg, err := p.manager.GetCgroups(); err != nil {
+			return fmt.Errorf("getting container cgroups: %w", err)
+		} else if cg.CpusetCpus != "" {
+			cpuAffinity, err = getCPUAffinity(os.DirFS("/"), cg.CpusetCpus)
+			if err != nil {
+				return fmt.Errorf("setting CPU affinity: %w", err)
+			}
+		}
+	}
+
+	err := startCommandWithCPUAffinity(p.cmd, cpuAffinity)
```

Comment:
I think maybe we should move the `cpuAffinity` check to here to make the code more readable:
```
	var err Error
	if cpuAffinity < 0 {
		err = cmd.Start()
	} else {
		err = startCommandWithCPUAffinity(p.cmd, cpuAffinity)
        }
```